feat: added logic to handle labeled answers

brianshen3 opened the pull request:
### Linked issue(s)

### What change does this PR introduce and why?
Handle compute metrics for labeled answers in the script.

### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

brianshen3 committed:
added logic to handle labeled answers

marko-polo-cheno commented:
With a partially labelled dataset and `uv run python3 retrieval_augmented_generation/upload_results.py query_decomp --dataset-name financebenchmarknew --evaluate`, I am seeing `kolena> Failed to import module for data type 'SPECIAL/TIMESTAMP'`.

Also seeing metrics indicate the correct page is retrieved, but the wrong document is retrieved which is contradictory.
<img width="1276" alt="image" src="https://github.com/user-attachments/assets/c85d625c-dcb0-4271-a990-50fc066a877e" />

brianshen3 commented:
Fixed the timestamp error. Was due to new data type needed to be imported as it was read in the download.

As for the metrics error. It was because originally there was only one doc. But with labelling it could be a list of docs. Updated the logic to if retrieved_docs is True then is_page can be true. Let me know if that logic makes sense @marko-polo-cheno

brianshen3 committed:
metrics fix

brianshen3 commented on kolena/_utils/datatypes.py:
```diff
@@ -99,7 +99,7 @@ def data_category_to_module_name(self) -> str:
             return "kolena.annotation"
         if self == DataCategory.SPECIAL:
-            return "kolena._experimental.data_type.special"
+            return "kolena._experimental.special_data_type"
```
@nankolena There were errors regarding the imports. I couldn't find `kolena._experimental.data_type.special`

I fixed it to the one aligned in the docs. Let me know if this is right track.

nankolena replied:
Ah good call, missed this when refactoring the module, thanks for picking this up! Once this is release we should bump the min kolena version in this example

nankolena replied:
Actually lemme make a separate change for this so we can cut a minor release off of it quickly

brianshen3 replied:
Ok I will move the change outside this PR

nankolena replied:
Cut [1.51.1 client version](https://github.com/kolenaIO/kolena/releases/tag/1.51.1) just now, release might take a bit of time. Could you rebase and update this example to `>=1.51.1` later?

brianshen3 replied:
No rush, its okay

brianshen3 replied:
Updated the versioning.

nankolena commented on kolena/_utils/datatypes.py:
```diff
@@ -116,7 +116,7 @@ def _get_data_type(name: str) -> Optional[Type["TypedDataObject"]]:
             importlib.import_module(module)
         except Exception as e:
-            log.error(f"Failed to import module for data type '{name}'", e)
+            log.error(f"Failed to2 import module for data type '{name}'", e)
```
nit - clean up

marko-polo-cheno commented on examples/dataset/retrieval_augmented_generation/retrieval_augmented_generation/metrics.py:
```diff
@@ -15,22 +15,47 @@
+            if x is not None
+            else [],
+        )
+        ground_truth_columns = ["doc_name", "relevant_pages", "financebench_id"]
```
Can we move `ground_truth_columns` outside of the if-statement?

Something like this might be cleaner
```
def extract_doc_names(labeling_task):
    if labeling_task is None:
        return []
    return [
        content.locator.split("/")[-1].replace(".pdf", "")
        for content in labeling_task.get("retrieved_contents", []) or []
    ]

def extract_relevant_pages(labeling_task):
    if labeling_task is None:
        return []
    return [
        content.page_number
        for content in labeling_task.get("retrieved_contents", []) or []
    ]

df_dataset["doc_name"] = df_dataset["labeling_task"].apply(extract_doc_names)
df_dataset["relevant_pages"] = df_dataset["labeling_task"].apply(extract_relevant_pages)

```

brianshen3 replied:
moved to cleaner logic. I also aligned to use doc_names instead of doc_name.

marko-polo-cheno commented on examples/dataset/retrieval_augmented_generation/retrieval_augmented_generation/metrics.py:
```diff
@@ -15,22 +15,47 @@
-    return any([doc_name in content.locator for content in retrieved_contents])
+def is_doc_retrieved(retrieved_contents: list, doc_names: list) -> bool:
+    # Handle case where doc_names is a list
+    return any(any(doc_name in content.locator for content in retrieved_contents) for doc_name in doc_names)
```
Should specify that we mark a model result as True/correct if just one document from the ground truth is found in the retrieved docs and likewise for the page metric.

In other words, any non-zero overlap for two sets of citations is treated as correct. Further improvements for the future would probably include an IOU score.

brianshen3 replied:
that's a good suggestion, I will update to reflect that

brianshen3 replied:
Updated to have note for the logic we use

brianshen3 committed:
updated is_page logic to be pair of doc + page_num. Added new versioning

brianshen3 committed:
added notes for rules on what is correct
